pandas_schema/schema.py

```diff
@@ -45,11 +45,51 @@ def validate(self, df: pd.DataFrame, columns: typing.List[str] = None) -> typing
             schema_cols = len(self.columns)
             columns_to_pair = self.columns
             if df_cols != schema_cols:
+
+                schema_columns = set(self.get_column_names())
+                df_columns = set(df.columns)
+
+                add_schema_columns = [col for col in schema_columns if col not in df_columns]
+                add_df_columns = [col for col in df_columns if col not in schema_columns]
+
+                if not add_schema_columns:
+
+                    errors.append(
+                        ValidationWarning(
+                            'Invalid number of columns. The schema specifies {n_schema}, '
+                            'but the data frame has {n_df}. '
+                            'The additional data frame columns are: {add_columns}.'.format(
+                                n_schema=schema_cols,
+                                n_df=df_cols,
+                                add_columns=add_df_columns,
+                            )
+                        )
+                    )
+                    return errors
+
+                if not add_df_columns:
+                    errors.append(
+                        ValidationWarning(
+                            'Invalid number of columns. The schema specifies {n_schema}, '
+                            'but the data frame has {n_df}. The additional schema columns are: {add_columns}.'.format(
+                                n_schema=schema_cols,
+                                n_df=df_cols,
+                                add_columns=add_schema_columns,
+                            )
+                        )
+                    )
+                    return errors
+
                 errors.append(
```

Comment:
Why do you have a second `ValidationWarning` added here? You already have one error if `add_schema_columns` and one error if `add_df_columns`. If there is an error with the number of columns this will produce two `ValidationWarning` objects, when it should produce one.